src/abstraction/ek.rs

```diff
@@ -0,0 +1,171 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    abstraction::nv,
+    constants::{
+        algorithm::{AsymmetricAlgorithm, Cipher, HashingAlgorithm},
+        tss::*,
+        types::session::SessionType,
+    },
+    handles::{AuthHandle, KeyHandle, NvIndexTpmHandle, TpmHandle},
+    session::Session,
+    tss2_esys::{
+        ESYS_TR_RH_ENDORSEMENT, TPM2B_ECC_PARAMETER, TPM2B_PUBLIC, TPM2B_PUBLIC_KEY_RSA,
+        TPMS_ECC_PARMS, TPMS_ECC_POINT, TPMS_RSA_PARMS, TPMS_SCHEME_HASH, TPMT_ECC_SCHEME,
+        TPMT_KDF_SCHEME, TPMT_RSA_SCHEME, TPMT_SYM_DEF_OBJECT, TPMU_ASYM_SCHEME, TPMU_SYM_KEY_BITS,
+        TPMU_SYM_MODE,
+    },
+    utils::{
+        ObjectAttributes, PublicIdUnion, PublicParmsUnion, Tpm2BPublicBuilder, TpmaSessionBuilder,
+    },
+    Context, Result,
+};
+
+// Source: TCG EK Credential Profile for TPM Family 2.0; Level 0 Version 2.3 Revision 2
+// Section 2.2.1.4 (Low Range) for Windows compatibility
+const RSA_2048_EK_CERTIFICATE_NV_INDEX: u32 = 0x01c00002;
+const ECC_P256_EK_CERTIFICATE_NV_INDEX: u32 = 0x01c0000a;
+
+fn create_nullauth_session(context: &mut Context) -> Result<Option<Session>> {
+    let session = context.start_auth_session(
+        None,
+        None,
+        None,
+        SessionType::Hmac,
+        Cipher::aes_256_cfb(),
+        HashingAlgorithm::Sha256,
+    )?;
+    let session_attr = TpmaSessionBuilder::new()
+        .with_flag(TPMA_SESSION_DECRYPT)
+        .with_flag(TPMA_SESSION_ENCRYPT)
+        .build();
+    context.tr_sess_set_attributes(session.unwrap(), session_attr)?;
+
+    Ok(session)
+}
+
+// Source: TCG EK Credential Profile for TPM Family 2.0; Level 0 Version 2.3 Revision 2
+// Appendix B.3.3 and B.3.4
+fn create_ek_public_from_default_template(alg: AsymmetricAlgorithm) -> Result<TPM2B_PUBLIC> {
+    let mut obj_attrs = ObjectAttributes(0);
+    obj_attrs.set_fixed_tpm(true);
+    obj_attrs.set_st_clear(false);
+    obj_attrs.set_fixed_parent(true);
+    obj_attrs.set_sensitive_data_origin(true);
+    obj_attrs.set_user_with_auth(false);
+    obj_attrs.set_admin_with_policy(true);
+    obj_attrs.set_no_da(false);
+    obj_attrs.set_encrypted_duplication(false);
+    obj_attrs.set_restricted(true);
+    obj_attrs.set_decrypt(true);
+    obj_attrs.set_sign_encrypt(false);
+
+    // TPM2_PolicySecret(TPM_RH_ENDORSEMENT)
+    // With 32 null-bytes attached, because of the type of with_auth_policy
+    let authpolicy: [u8; 64] = [
+        0x83, 0x71, 0x97, 0x67, 0x44, 0x84, 0xb3, 0xf8, 0x1a, 0x90, 0xcc, 0x8d, 0x46, 0xa5, 0xd7,
+        0x24, 0xfd, 0x52, 0xd7, 0x6e, 0x06, 0x52, 0x0b, 0x64, 0xf2, 0xa1, 0xda, 0x1b, 0x33, 0x14,
+        0x69, 0xaa, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00,
+    ];
```

Comment:
How was this computed?

Reply:
That is the TPM policy digest value of a clear session where `TPM2_PolicySecret(TPM_RH_ENDORSEMENT)` was executed (theoretically).
This value comes straight from the TCG EK Credential Profile for TPM Family 2.0; Level 0 Version 2.3 Revision 2 spec.